test/distributed/fsdp/test_fsdp_optim_state.py

```diff
@@ -0,0 +1,435 @@
+                {"params": model.param_group0()},
+                {"params": model.param_group1(), "weight_decay": 0.9}
+            ]
+        optim = optim_class(optim_input, lr=lr, **optim_kwargs)
+        return model, optim, optim_input
+
+    def _init_transformer_model(
+        self,
+        wrap: bool,
+        device: torch.device = torch.device("cuda"),
+        group=None,
+        optim_class: Type[torch.optim.Optimizer] = torch.optim.Adam,
+        use_multiple_param_groups: bool = False,
+        **optim_kwargs,
+    ):
+        assert not use_multiple_param_groups, \
+            "Multiple parameter groups for the transformer is not implemented"
+        if group is None:
+            group = dist.distributed_c10d._get_default_group()
+        model = self._get_wrapped_model(group=group).to(device) if wrap \
+            else self._get_nonwrapped_model(group=group).to(device)
+        model.eval()  # disable dropout for determinism
+        lr = optim_kwargs.pop("lr", 0.01)
+        optim = optim_class(model.parameters(), lr=lr, **optim_kwargs)
+        return model, optim, None
+
+    def _step_model(
+        self,
+        model: torch.nn.Module,
+        optim: torch.optim.Optimizer,
+        device: torch.device = torch.device("cuda"),
+        num_iters: int = 1,
+    ) -> List[float]:
+        """Performs a forward pass, backward pass, and optimizer step
+        ``num_iters``-many times, and returns the per-iteration losses."""
+        torch.manual_seed(0)  # set seed for determinism
+        losses = []
+        module = model.module if hasattr(model, "module") else model
+        for _ in range(num_iters):
+            inp = module.get_input(device)
+            output = model(*inp)
+            loss = module.get_loss(inp, output).to(device)
+            losses.append(loss.item())
+            module.run_backward(loss)
+            optim.step()
+        return losses
+
+    def _broadcast_full_osd(self, full_osd: Dict[str, Any], group=None):
+        """Broadcasts the full optimizer state dict in place of using
+        ``torch.save()`` and ``torch.load()`` so that all ranks can have it."""
+        obj_list = [full_osd]
+        dist.broadcast_object_list(
+            obj_list, src=OPTIM_TARGET_RANK, group=group,
+        )
+        full_osd = obj_list[0]
+        return full_osd
+
+    def _are_equal_states(
+        self,
+        state1: Dict[str, Any],
+        state2: Dict[str, Any],
+    ) -> bool:
+        """Checks if ``state1`` and ``state2`` contain the same mappings."""
+        if set(state1.keys()) != set(state2.keys()):
+            return False
+        for state_name, value1 in state1.items():
+            value2 = state2[state_name]
+            if type(value1) != type(value2):
+                return False
+            if torch.is_tensor(value1):  # tensor state
+                assert torch.is_tensor(value2)
+                # Check the values on CPU to be device-agnostic
```

Comment:
We have not yet settled on what the right device semantics are, so I am leaving the test case device agonistic.